src/end_entity.rs

```diff
@@ -185,4 +194,86 @@ impl<'a> EndEntityCert<'a> {
+    /// If the provided key isn't usable for this EE certificate [`Error::CertPrivateKeyMismatch`]
+    /// will be returned.
+    ///
+    /// This function supports the following private key algorithms and encodings (matching the
+    /// supported formats used by Rustls):
+    /// Key algorithms:
+    ///   * RSA
+    ///   * ECDSA (P-256 or P-384)
+    ///   * Ed25519
+    /// Encodings:
+    ///   * PKCS8v1 (RSA, ECDSA, Ed25519)
+    ///   * PKCS8v2 (Ed25519 only)
+    ///   * PKCS1 (RSA only)
+    ///   * Sec1 (ECDSA only)
+    pub fn verify_private_key(&self, private_key_der: &[u8]) -> Result<(), Error> {
+        // Parse the SPKI of the EE cert and extract the DER encoded bytes of the public key.
+        let spki = parse_spki_value(self.inner.spki.value())?;
+        let pubkey = spki.key_value.as_slice_less_safe();
+
+        fn matches_or_error(pubkey: &[u8], pubkey_tbv: &[u8]) -> Result<(), Error> {
+            match pubkey == pubkey_tbv {
+                true => Ok(()),
+                false => Err(Error::CertPrivateKeyMismatch),
+            }
+        }
+
+        #[cfg(feature = "alloc")]
+        if let Some(rsa_keypair) = Self::rsa_from_der(private_key_der) {
+            return matches_or_error(pubkey, rsa_keypair.public_key().as_ref());
+        }
+
+        if let Some(ecdsa_keypair) = Self::ecdsa_from_pkcs8(private_key_der) {
+            return matches_or_error(pubkey, ecdsa_keypair.public_key().as_ref());
+        }
+
+        #[cfg(feature = "alloc")]
+        if let Some(ecdsa_keypair) = Self::ecdsa_from_sec1(private_key_der) {
+            return matches_or_error(pubkey, ecdsa_keypair.public_key().as_ref());
+        }
+
+        if let Some(ed25519_keypair) = Self::ed25519_from_pkcs8(private_key_der) {
+            return matches_or_error(pubkey, ed25519_keypair.public_key().as_ref());
+        }
+
+        Err(Error::CertPrivateKeyMismatch)
+    }
+
+    #[cfg(feature = "alloc")]
+    fn rsa_from_der(private_key_der: &[u8]) -> Option<RsaKeyPair> {
+        RsaKeyPair::from_pkcs8(private_key_der)
+            .or_else(|_| RsaKeyPair::from_der(private_key_der))
+            .ok()
+    }
+
+    fn ecdsa_from_pkcs8(pkcs8_der: &[u8]) -> Option<EcdsaKeyPair> {
+        EcdsaKeyPair::from_pkcs8(&ECDSA_P256_SHA256_ASN1_SIGNING, pkcs8_der)
+            .ok()
+            .or_else(|| EcdsaKeyPair::from_pkcs8(&ECDSA_P384_SHA384_ASN1_SIGNING, pkcs8_der).ok())
+    }
+
+    #[cfg(feature = "alloc")]
+    fn ecdsa_from_sec1(sec1_der: &[u8]) -> Option<EcdsaKeyPair> {
+        fn try_sec1_curve(
+            sigalg: &'static EcdsaSigningAlgorithm,
+            sec1_der: &[u8],
+        ) -> Option<EcdsaKeyPair> {
+            der::convert_sec1_to_pkcs8(sigalg, sec1_der)
+                .and_then(|pkcs8_der| {
+                    EcdsaKeyPair::from_pkcs8(sigalg, &pkcs8_der).map_err(|_| Error::BadDer)
+                })
+                .ok()
+        }
```

Comment:
I think this would be easier to follow if you extracted the nested function.

Reply:
Done.